File: adventjs/2025/21/solution.py
def clear_gifts(warehouse: list[list[str]], drops: list[int]) -> list[list[str]]:
  result = [row[:] for row in warehouse]
  cols = len(warehouse[0])
  
  for col in drops:
    drop_row = -1
    for row in range(len(result) - 1, -1, -1):
      if result[row][col] == '.':
        drop_row = row
        break
    
    if drop_row == -1:
      continue
    
    result[drop_row][col] = '#'
    
    if all(cell == '#' for cell in result[drop_row]):
      result.pop(drop_row)
      result.insert(0, ['.'] * cols)
  
  return result

File: adventjs/2025/21/solution.js
/**
 * @param {string[][]} warehouse
 * @param {number[]} drops
 * @returns {string[][]}
 */
function clearGifts(warehouse, drops) {
  const result = warehouse.map(row => [...row]);
  const cols = warehouse[0].length;
  
  for (const col of drops) {
    let dropRow = -1;
    for (let row = result.length - 1; row >= 0; row--) {
      if (result[row][col] === '.') {
        dropRow = row;
        break;
      }
    }
    
    if (dropRow === -1) continue;
    
    result[dropRow][col] = '#';
    
    if (result[dropRow].every(cell => cell === '#')) {
      result.splice(dropRow, 1);
      result.unshift(new Array(cols).fill('.'));
    }
  }
  
  return result;
}
